core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,292 @@
+    activeBuffer = ByteBuffer.allocate(bufferSizeInBytes);
+    readAheadBuffer = ByteBuffer.allocate(bufferSizeInBytes);
+    this.readAheadThresholdInBytes = readAheadThresholdInBytes;
+    this.underlyingInputStream = inputStream;
+    activeBuffer.flip();
+    readAheadBuffer.flip();
+  }
+
+  private boolean isEndOfStream() {
+    if(activeBuffer.remaining() == 0 && readAheadBuffer.remaining() == 0 && endOfStream) {
+      return true;
+    }
+    return  false;
+  }
+
+
+  private void readAsync(final ByteBuffer byteBuffer) throws IOException {
+    stateChangeLock.lock();
+    if (endOfStream || isReadInProgress) {
+      stateChangeLock.unlock();
+      return;
+    }
+    byteBuffer.position(0);
+    byteBuffer.flip();
+    isReadInProgress = true;
+    stateChangeLock.unlock();
+    executorService.execute(() -> {
+      byte[] arr;
+      stateChangeLock.lock();
+      arr = byteBuffer.array();
+      stateChangeLock.unlock();
+      // Please note that it is safe to release the lock and read into the read ahead buffer
+      // because either of following two conditions will hold - 1. The active buffer has
+      // data available to read so the reader will not read from the read ahead buffer.
+      // 2. This is the first time read is called or the active buffer is exhausted,
+      // in that case the reader waits for this async read to complete.
+      // So there is no race condition in both the situations.
+      int nRead = 0;
+      while (nRead == 0) {
+        try {
+          nRead = underlyingInputStream.read(arr);
+          if (nRead < 0) {
+            // We hit end of the underlying input stream
+            break;
+          }
+        } catch (Exception e) {
+          stateChangeLock.lock();
+          // We hit a read exception, which should be propagated to the reader
+          // in the next read() call.
+          isReadAborted = true;
+          readException = e;
+          stateChangeLock.unlock();
+        }
+      }
+      stateChangeLock.lock();
+      if (nRead < 0) {
+        endOfStream = true;
+      }
+      else {
+        // fill the byte buffer
+        byteBuffer.limit(nRead);
+      }
+      isReadInProgress = false;
+      signalAsyncReadComplete();
+      stateChangeLock.unlock();
+    });
+  }
+
+  private void signalAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.signalAll();
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  private void waitForAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.await();
+    } catch (InterruptedException e) {
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    int val = read(oneByte, 0, 1);
+    if (val == -1) {
+      return -1;
+    }
+    return oneByte[0] & 0xFF;
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
```

Comment:
@mridulm - This is synchronized because the stateChangeLock can be reliniquished by any read call in the waitForAsyncReadComplete function. so depending on stateChangeLock for synchronizing external apis are not thread-safe.

Reply:
As I mention above, as long as state is read/updated only with the lock held, I dont see how release of lock in `waitForAsyncReadComplete` is causing issues.
`stateChangeLock` is a reentrant lock, so if at entry lock was held, on exit it will continue to be held; and if at entry it was not, at exit it will not be due to the try/finally acquired/release [1].
Did I miss something here ? Perhaps I misunderstood your comment ?

[1]
```
import java.util.concurrent.locks.ReentrantLock
val stateChangeLock = new ReentrantLock()
stateChangeLock.lock()
stateChangeLock.lock()
stateChangeLock.unlock()
stateChangeLock.isLocked
stateChangeLock.unlock()
stateChangeLock.isLocked
```

Reply:
There are few reasons I am hesitant to remove the synchronized from the public apis.

1. Some variables like executorService, underlyingInputStream are not protected by the statechange lock. There might be a race condition when we try to execute executorService.execute call at the same time shutdown the executor service in close function call.
2. I am not sure what will be the behavior if multiple threads are waiting on  asyncReadComplete.await() call. Is the signaling thread going to wake only one of them? If yes, is it going to do it in the order? If it is not guaranteed to wake up the callers in order, then the read calls which are issued from different threads at different point of time are not guaranteed to read them in order. Is that acceptable?

Reply:
`executorService` and `underlyingInputStream` are `final` right ? So there is no concurrent mods to them.
If you are worried about thread safety of these - `executorService` is thread safe (jdk contract).
In your code, `underlyingInputStream` must be invoked in order (if it is not, then we have a bug) : `synchronized` block is not protecting `underlyingInputStream` anyway.


For reads on `ReadAheadInputStream`: there is no gaurantee of ordering in case of parallel read's for InputStream's - did I miss some expectation that this class provides that gaurantee to its consumers ?

Note that I want to understand why we need `synchronized`, and if there is a need for it, sure we can keep it as such :-)
The only reason I see, as of now, is that use of `oneByte` array is problematic : if we eliminate that via a thread local or rewrite code to handle `read()` without using `read(byte[])`, then I dont see a current need for it. Please let me know if I am missing something.

Reply:
@mridulm - After iterating through the code several times and thinking about it I convinced myself that we do not need the `synchronized` for the public apis.  I also changed the use the `oneByte` array to be thread local so that we do not have any race-condition there.

Reply:
Thanks for clarifying @sitalkedia.
I was not sure if I was missing something, particularly for the way it is being used and the asyncRead vs normal read usecases, w.r.t synchronization requirements.

I will wait for @zsxwing to also take a look, but this is looking really good !